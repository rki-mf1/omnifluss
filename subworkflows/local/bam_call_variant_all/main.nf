include { LOFREQ_VITERBI }      from '../../../modules/nf-core/lofreq/viterbi/main'
include { LOFREQ_INDELQUAL }    from '../../../modules/nf-core/lofreq/indelqual/main'
include { LOFREQ_CALLPARALLEL } from '../../../modules/nf-core/lofreq/callparallel/main'
include { SAMTOOLS_INDEX }      from '../../../modules/nf-core/samtools/index/main'
include { BCFTOOLS_NORM }       from '../../../modules/nf-core/bcftools/norm/main'
include { BCFTOOLS_INDEX}       from '../../../modules/nf-core/bcftools/index/main'


workflow BAM_CALL_VARIANT_ALL {

    take:
    tools           // String
    ch_bam          // channel: [ val(meta), [ bam ]   ]
    ch_ref          // channel: [ val(meta), [ fasta ] ]
    ch_ref_index    // channel: [ val(meta), [ fai ]   ]


    main:
    ch_corrbam      = Channel.empty()
    ch_iqbam        = Channel.empty()
    ch_iqbam_bai    = Channel.empty()
    ch_vcf          = Channel.empty()
    ch_tbi          = Channel.empty()
    ch_versions     = Channel.empty()

    /****************************************************************/
    /* STEP 0: Format input Channels for sample-wise ordered tuples */
    /****************************************************************/
    ch_bam_cpy          = ch_bam.map{ meta, bam -> return [meta.id, meta, bam ] }
    ch_ref_cpy          = ch_ref.map{ meta, fasta -> return [meta.id, meta, fasta] }
    ch_ref_index_cpy    = ch_ref_index.map{ meta, fai -> return [meta.id, meta, fai] }

    ch_lofreq_viterbi_input = ch_bam_cpy.join(ch_ref_cpy).join(ch_ref_index_cpy)
        .multiMap{_sample_id, meta, bam, meta2, fasta, meta3, fai ->
            ch_bam: [meta, bam]
            ch_ref: [meta2, fasta]
            ch_ref_index: [meta3, fai]
        }

    if (tools.split(',').contains('lofreq')) {
        /****************************************************************/
        /* STEP 1: LOFREQ_VITERBI for mapping-error correction          */
        /****************************************************************/
        // LOFREQ_VITERBI nf-core module was implemented by @MarieLataretu and includes samtools sort. No need for an extra sorting step here.
        LOFREQ_VITERBI (
            ch_lofreq_viterbi_input.ch_bam,
            ch_lofreq_viterbi_input.ch_ref
        )
        ch_corrbam  = LOFREQ_VITERBI.out.bam
        ch_versions = ch_versions.mix(LOFREQ_VITERBI.out.versions.first())

        /*******************************************************************/
        /* STEP 2: LOFREQ_INDELQUAL adds indel qualities to the alignments */
        /*******************************************************************/
        ch_corrbam_cpy  = ch_corrbam.map{ meta, corrbam -> return [meta.id, meta, corrbam ] }

        ch_lofreq_indelqual_input = ch_corrbam_cpy.join(ch_ref_cpy)
            .multiMap{_sample_id, meta, corrbam, meta2, fasta ->
                ch_corrbam: [meta, corrbam]
                ch_ref: [meta2, fasta]
            }

        LOFREQ_INDELQUAL (
            ch_lofreq_indelqual_input.ch_corrbam,
            ch_lofreq_indelqual_input.ch_ref
        )
        ch_iqbam    = LOFREQ_INDELQUAL.out.bam
        ch_versions = ch_versions.mix(LOFREQ_INDELQUAL.out.versions.first())

        // LOFREQ_ALNQUAL adds alignment quality per base and indel
        // LOFREQ_ALNQUAL will not be taken over from FluPipe for now, as it is computed on the fly in lofreq call/call-parallel

        /*******************************************************************/
        /* STEP 3: SAMTOOLS_INDEX creates the bam index (bai)              */
        /*******************************************************************/
        SAMTOOLS_INDEX (
            ch_iqbam
        )
        ch_iqbam_bai    = SAMTOOLS_INDEX.out.bai    // [ val(meta), path("*.bai") ]
        ch_versions     = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())

        /*******************************************************************/
        /* STEP 4: LOFREQ_CALLPARALLEL does the actual variant calling     */
        /*******************************************************************/
        ch_iqbam_cpy        = ch_iqbam.map{ meta, iqbam -> return [meta.id, meta, iqbam ] }
        ch_iqbam_bai_cpy    = ch_iqbam_bai.map{ meta, iqbai -> return [meta.id, meta, iqbai ] }

        ch_lofreq_callparallel_input = ch_iqbam_cpy.join(ch_iqbam_bai_cpy).join(ch_ref_cpy).join(ch_ref_index_cpy)
            .multiMap{ _sample_id, meta, iqbam, _meta2, iqbai, meta3, fasta, meta4, fai ->
                ch_input_sample: [meta, iqbam, iqbai, [] ]  // needs empty optional parameter here
                ch_ref: [meta3, fasta]
                ch_ref_index: [meta4, fai]
            }

        LOFREQ_CALLPARALLEL (
            ch_lofreq_callparallel_input.ch_input_sample,
            ch_lofreq_callparallel_input.ch_ref,
            ch_lofreq_callparallel_input.ch_ref_index
        )
        ch_vcf      = LOFREQ_CALLPARALLEL.out.vcf
        ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())
    }

    /*******************************************************************/
    /* STEP 5: Caller-independent postprocessing                       */
    /* STEP 5: BCFTOOLS_INDEX createse an index of the VCF file        */
    /*******************************************************************/
    BCFTOOLS_INDEX (
        ch_vcf
    )
    ch_tbi      = BCFTOOLS_INDEX.out.tbi
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX.out.versions.first())

    /**********************************************************************/
    /* STEP 6: BCFTOOLS_NORM normalizes variant notations and positioning */
    /**********************************************************************/
    ch_vcf_cpy = ch_vcf.map{ meta, vcf -> return [meta.id, meta, vcf ] }
    ch_tbi_cpy = ch_tbi.map{ meta, tbi -> return [meta.id, meta, tbi ] }

    ch_bcftools_norm_input = ch_vcf_cpy.join(ch_tbi_cpy).join(ch_ref_cpy)
        .multiMap{_sample_id, meta, vcf, _meta2, tbi, meta3, fasta ->
            ch_input_variants: [meta, vcf, tbi]
            ch_ref: [meta3, fasta]
        }

    // Normalize variant notations and positioning; index normed VCF
    BCFTOOLS_NORM(
        ch_bcftools_norm_input.ch_input_variants,
        ch_bcftools_norm_input.ch_ref
    )
    ch_vcf      = BCFTOOLS_NORM.out.vcf     // overwrite VCF with VCF after normalization
    ch_tbi      = BCFTOOLS_NORM.out.tbi     // overwrite TBI with TBI after normalization
    ch_versions = ch_versions.mix(BCFTOOLS_NORM.out.versions.first())


    emit:
    bam         = ch_iqbam      // channel: [ val(meta), [ bam ] ]  // Not the same as input bam! Here, the bam got normalized and annotated.
    bai         = ch_iqbam_bai  // channel: [ val(meta), [ bai ] ]
    vcf         = ch_vcf        // channel: [ val(meta), [ vcf ] ]
    tbi         = ch_tbi        // channel: [ val(meta), [ tbi ] ]
    versions    = ch_versions   // channel: [ versions.yml ]
}

