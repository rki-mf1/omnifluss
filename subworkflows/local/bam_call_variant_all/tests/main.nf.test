nextflow_workflow {

    name "Test Subworkflow BAM_CALL_VARIANT_ALL"
    script "../main.nf"
    // Incude a nextflow.config for the test if needed
    config "subworkflows/local/bam_call_variant_all/tests/nextflow.config"
    workflow "BAM_CALL_VARIANT_ALL"

    //  Add tags for the subworkflow
    tag "subworkflows"
    tag "subworkflows/local"
    tag "subworkflows/bam_call_variant_all"
    tag "bam_call_variant_all"
    // Add tags for all modules used within this subworkflow.
    tag "lofreq/viterbi"
    tag "lofreq/indelqual"
    tag "lofreq/callparallel"
    tag "samtools/index"


    // Test name preferably indicating the test-data and file-format used
    test("sarscov2 - bam") {

        // Activate the stub mode for nextflow
        //options '-stub'

        when {
            workflow {
                """
                input[0] = 'lofreq'
                input[1] = [ [ id:'test', single_end:false ], // meta map
                            [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true) ]
                        ]
                input[2] = file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                input[3] = file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}