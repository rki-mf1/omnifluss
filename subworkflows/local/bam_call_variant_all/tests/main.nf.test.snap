{
    "sarscov2 - bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.indelqual.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.indelqual.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.normed.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.normed.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,03d3f12fc45600a49531ffdca6340ec6",
                    "versions.yml:md5,2bbb3ad61356fe366ce767350d36b299",
                    "versions.yml:md5,3ee8176c6b98480d5a06b8d0e5fc2f37",
                    "versions.yml:md5,a3aef5bda80ccafb31e8aad82ae46fb5",
                    "versions.yml:md5,a3aef5bda80ccafb31e8aad82ae46fb5",
                    "versions.yml:md5,ddbf5164dc5b4349c6f71618b3bf24f9"
                ],
                "bai": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.indelqual.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.indelqual.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.normed.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.normed.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,03d3f12fc45600a49531ffdca6340ec6",
                    "versions.yml:md5,2bbb3ad61356fe366ce767350d36b299",
                    "versions.yml:md5,3ee8176c6b98480d5a06b8d0e5fc2f37",
                    "versions.yml:md5,a3aef5bda80ccafb31e8aad82ae46fb5",
                    "versions.yml:md5,a3aef5bda80ccafb31e8aad82ae46fb5",
                    "versions.yml:md5,ddbf5164dc5b4349c6f71618b3bf24f9"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.1"
        },
        "timestamp": "2025-03-07T11:13:05.475766545"
    },
    "sarscov2 - bam": {
        "content": [
            [
                "##INFO=<ID=AF,Number=1,Type=Float,Description=\"Allele Frequency\">",
                "##INFO=<ID=SB,Number=1,Type=Integer,Description=\"Phred-scaled strand bias at this position\">",
                "##INFO=<ID=DP4,Number=4,Type=Integer,Description=\"Counts for ref-forward bases, ref-reverse, alt-forward and alt-reverse bases\">",
                "##INFO=<ID=INDEL,Number=0,Type=Flag,Description=\"Indicates that the variant is an INDEL.\">",
                "##INFO=<ID=CONSVAR,Number=0,Type=Flag,Description=\"Indicates that the variant is a consensus variant (as opposed to a low frequency variant).\">",
                "##INFO=<ID=HRUN,Number=1,Type=Integer,Description=\"Homopolymer length to the right of report indel position\">",
                "##FILTER=<ID=min_snvqual_31,Description=\"Minimum SNV Quality (Phred) 31\">",
                "##FILTER=<ID=min_indelqual_20,Description=\"Minimum Indel Quality (Phred) 20\">",
                "##FILTER=<ID=min_dp_10,Description=\"Minimum Coverage 10\">",
                "##FILTER=<ID=sb_fdr,Description=\"Strand-Bias Multiple Testing Correction: fdr corr. pvalue > 0.001000\">",
                "##FILTER=<ID=min_snvqual_43,Description=\"Minimum SNV Quality (Phred) 43\">",
                "##FILTER=<ID=min_indelqual_28,Description=\"Minimum Indel Quality (Phred) 28\">",
                "##bcftools_normVersion=1.20+htslib-1.20"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.1"
        },
        "timestamp": "2025-03-07T11:12:54.662479838"
    }
}